File: oop/lab1/cpp/src/sieve.cpp
#include <sieve.hpp>

#include <cstddef>
#include <stdexcept>
#include <vector>
#include <string>

std::vector<int> Sieve::Primes(int n) noexcept(false)
{
    if (n < 2)
    {
        throw std::invalid_argument("parameter n should be >= 2, got " + std::to_string(n));
    }

    std::vector<int> primes;
    std::vector<bool> isPrime(static_cast<size_t>(n + 1), true);

    for (size_t i = 2; i <= static_cast<size_t>(n) ; i++)
    {
        if (isPrime[i])
        {
            primes.push_back(static_cast<int>(i));
            for (size_t j = i * i; j <= static_cast<size_t>(n); j += i)
            {
                isPrime[j] = false;
            }
        }
    }

    return primes;
}

File: 2024_semester_2/programming-course/lab1/java/Sieve.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class Sieve {
    private Sieve() throws InstantiationError {
        throw new InstantiationError("This is a static class!");
    }

    public static List<Integer> primes(int n) throws IllegalArgumentException {
        if (n < 2) {
            throw new IllegalArgumentException("Parameter n should be >= 2, got " + n);
        }

        boolean[] isPrime = new boolean[n + 1];
        Arrays.fill(isPrime, true);
        List<Integer> primes = new ArrayList<>();

        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                primes.add(i);
                for (long j = (long) i * i; j <= (long)n; j += i) {
                    isPrime[(int) j] = false;
                }
            }
        }

        return primes;
    }
}
